# 전처리 & 머신러닝 통합 개인과제

## Part 1. 고객 데이터 품질 개선

### 실무 시나리오

전자상거래 기업의 CRM팀은 고객별 구매 패턴을 분석하고 구매 가능성이 높은 고객을 선별하려고 합니다.  
그러나 전달받은 원본 고객 데이터에는 잘못된 자료형, 범주 표기 불일치, 결측치, 중복 데이터와 극단값이 포함되어 있어 바로 분석에 사용할 수 없습니다.

CRM팀은 **데이터 분석 담당자**로서 원본 데이터의 품질 문제를 진단하고, 이후 Part 2의 EDA와 Part 3의 구매 예측 모델링에 사용할 수 있는 분석용 데이터셋을 만들어야 합니다.

### 과제 목표

- 데이터의 구조·자료형·기초 분포를 확인하고 주요 품질 문제를 파악합니다.
- 결측치, 중복값, 이상치와 범주 표기 불일치를 분석 목적에 맞게 처리합니다.
- 조건 기반 데이터 추출과 파생변수 생성을 수행합니다.
- 전처리 결과를 검증하고 다음 Part에서 사용할 CSV 파일로 저장합니다.

### 사용 환경 및 데이터

- Python 3.X
- pandas, numpy
- `전자상거래_고객구매_원본데이터.csv`

### 제출 결과물

- Part 1 실습 노트북
- `전자상거래_고객구매_전처리완료.csv`

> 문제 1~3은 필수 문제입니다.

## 문제 1. 원본 고객 데이터 품질 진단

### 업무 상황

CRM팀에 분석 일정을 공유하기 전에 원본 데이터가 실제 분석에 사용할 수 있는 상태인지 확인해야 합니다.  
데이터를 수정하기 전에 구조와 품질 문제를 먼저 점검하고, 이후 처리해야 할 항목을 정리하세요.

### 요구사항

1. 원본 데이터를 불러오고 데이터의 크기, 컬럼, 자료형과 기초 통계량을 확인하세요.
2. 컬럼별 결측치와 전체 행 기준 중복 데이터를 확인하세요.
3. 주요 범주형 컬럼의 값을 확인하여 표기 불일치 여부를 파악하세요.
4. 이후 정제가 필요한 주요 품질 문제를 간단히 정리하세요.

### 힌트

- 데이터 구조와 기초 통계량을 함께 확인하면 자료형 오류와 비정상 범위를 찾기 쉽습니다.
- 범주형 컬럼은 고유값을 확인하여 대소문자, 공백, 한글·영문 혼용 여부를 살펴볼 수 있습니다.

In [2]:
import pandas as pd
import numpy as np
from pathlib import Path

DATA_PATH = Path("전자상거래_고객구매_원본데이터.csv")

# TODO: 원본 데이터를 불러와 df에 저장하세요.
df = pd.read_csv(DATA_PATH)

# TODO: 데이터의 크기, 컬럼, 자료형과 기초 통계량을 확인하세요.
print("데이터 크기:", df.shape)
print("컬럼:", df.columns.tolist())
# print("자료형:", df.dtypes())
print(df.describe())
print(df.value_counts())
# print(df["CustomerID"].unique())
# print(df["Gender"].unique())
# print(df["Age"].unique())
# print(df["Region"].unique())
# print(df["MembershipLevel"].unique())
# print(df["VisitCount"].unique())
# print(df["AveragePurchaseAmount"].unique())
# print(df["TotalPurchaseAmount"].unique())
# print(df["SatisfactionScore"].unique())
# print(df["LastPurchaseDate"].unique())
# print(df["CouponUsed"].unique())
# print(df["PreferredCategory"].unique())
# print(df["SignupDate"].unique())
# print(df["Email"].unique())
# print(df["PurchaseStatus"].unique())

# # TODO: 결측치, 중복 데이터와 주요 범주형 컬럼의 값을 확인하세요.
print("결측치:", df.isnull().sum())
print("중복데이터:", df.duplicated().sum())

# # 카테고리별 주문 횟수
# print(df.value_counts())
# df['Age'].nunique()

데이터 크기: (1560, 15)
컬럼: ['CustomerID', 'Gender', 'Age', 'Region', 'MembershipLevel', 'VisitCount', 'AveragePurchaseAmount', 'TotalPurchaseAmount', 'SatisfactionScore', 'LastPurchaseDate', 'CouponUsed', 'PreferredCategory', 'SignupDate', 'Email', 'PurchaseStatus']
        VisitCount  TotalPurchaseAmount  SatisfactionScore  PurchaseStatus
count  1560.000000         1.560000e+03        1521.000000     1560.000000
mean      9.521795         4.841402e+05           3.662853        0.485256
std       8.146008         1.410879e+06           0.759151        0.499943
min       1.000000         5.000000e+03           1.100000        0.000000
25%       7.000000         1.869500e+05           3.200000        0.000000
50%       9.000000         3.366000e+05           3.700000        0.000000
75%      11.000000         5.370750e+05           4.200000        1.000000
max     170.000000         3.000000e+07           5.000000        1.000000
CustomerID  Gender  Age  Region      MembershipLevel  VisitCou

### 품질 진단 결과

- 자료형 정리가 필요한 컬럼:Age-세, ?, unknown, 비어있음 등 문자열 포함/ 
AveragePurchaseAmount 문자열, ',' / 
LastPurchaseDate, SignupDate -날짜인데 문자열로 되어있음 / 
- 표기 통일이 필요한 컬럼:Age(세, ?, unknown, 비어있음 등)/
Gender(nan, 대소문자, other, F, M)/ Region(nan, 띄어쓰기, 대소문자) / MembershipLevel(대소문자, nan, 띄어쓰기) / CouponUsed(대소문자, no, yes, y, n 통일필요) / PreferredCategory(대소문자, 띄어쓰기)
- 결측치가 있는 컬럼: Gender, Age, Region, MembershipLevel, AveragePurchaseAmount,
SatisfactionScore(nan), LastPurchaseDate, PreferredCategory
- 중복 데이터 확인 결과: 60
- 이상치 확인이 필요한 컬럼: Age, VisitCount,  TotalPurchaseAmount
- 이후 처리할 주요 항목: 문자열을 수치형으로 변환, 날짜형식 통일, 이상치, 결측치, 중복 

## 문제 2. 분석 가능한 고객 데이터로 정제

### 업무 상황

품질 진단 결과를 바탕으로 고객 데이터를 분석 가능한 상태로 정리해야 합니다.  
처리 과정에서 고객의 실제 구매 행동 정보가 불필요하게 손실되지 않도록 데이터 특성을 고려하세요.

### 요구사항

1. 원본 데이터를 보존한 상태에서 정제용 데이터를 생성하세요.
2. 분석에 맞지 않는 수치형·날짜형 자료형과 범주 표기를 정리하세요.
3. 결측치와 전체 행 기준 중복 데이터를 처리하세요.
4. 주요 수치형 컬럼의 이상치를 탐지하고 적절한 방법으로 처리하세요.
5. 처리 전후의 결측치, 중복값과 이상치 상태를 확인하세요.

### 힌트

- 변환할 수 없는 문자열은 결측치로 바꾼 뒤 일관되게 처리할 수 있습니다.
- 이상치는 무조건 삭제하기보다 실제 우수 고객의 행동일 가능성도 고려하세요.
- IQR은 이상치 후보를 확인하는 대표적인 방법입니다.

In [3]:
# TODO: 원본을 보존하고 정제용 데이터 df_clean을 생성하세요.
df_clean = df.copy()
# TODO: 수치형·날짜형 자료형과 범주 표기를 정리하세요.
cols = ['Gender','Region','MembershipLevel','PreferredCategory','CouponUsed']
for col in cols:
    df_clean[col]=df_clean[col].astype(str).str.strip().str.title()
    df_clean[col] = df_clean[col].replace(["Nan", "nan", "NaN", "None", ""], np.nan)
    
df_clean['Age']= df_clean['Age'].astype(str).str.replace('세','',regex=False)
df_clean['Age'] = pd.to_numeric(df_clean['Age'], errors='coerce')

df_clean['MembershipLevel'] = df_clean['MembershipLevel'].replace({'Vip': 'VIP'})

df_clean['AveragePurchaseAmount']= df_clean['AveragePurchaseAmount'].astype(str).str.replace(',','',regex=False)
df_clean['AveragePurchaseAmount'] = pd.to_numeric(df_clean['AveragePurchaseAmount'], errors='coerce')

df_clean['LastPurchaseDate']= df_clean['LastPurchaseDate'].astype(str)
df_clean['LastPurchaseDate'] = df_clean['LastPurchaseDate'].str.replace('[./]','-',regex=True)
df_clean['LastPurchaseDate'] = pd.to_datetime(df_clean['LastPurchaseDate'], errors='coerce')

signup= df_clean['SignupDate'].astype(str).str.strip()
d1 = pd.to_datetime(signup, format="%Y-%m-%d", errors='coerce')
d2 = pd.to_datetime(signup, format= "%d-%m-%Y", errors='coerce')
df_clean['SignupDate'] = d1.fillna(d2)

df_clean['Gender']= df_clean['Gender'].replace({
    'M':'Male', 'f':'Female', 'F':'Female'
})
df_clean['CouponUsed']=df_clean['CouponUsed'].replace({
    'No':'N','no':'N', 'yes':'Y', 'Yes':'Y', 'YES':'Y'
})

df_clean['VisitCount']= pd.to_numeric(df_clean['VisitCount'], errors='coerce')


In [4]:
# print(df_clean["LastPurchaseDate"].dtype)
# print("결측치:", df_clean["LastPurchaseDate"].isnull().sum())
# print(df_clean["LastPurchaseDate"].unique())
# print(df.dtypes)
print("결측치:", df_clean.isnull().sum())

결측치: CustomerID                0
Gender                   26
Age                      48
Region                   31
MembershipLevel          22
VisitCount                0
AveragePurchaseAmount    45
TotalPurchaseAmount       0
SatisfactionScore        39
LastPurchaseDate         30
CouponUsed                0
PreferredCategory        26
SignupDate                0
Email                     0
PurchaseStatus            0
dtype: int64


In [5]:
# print("원본 결측치:", df["LastPurchaseDate"].isnull().sum())
# print("변환 후 결측치:", df_clean["LastPurchaseDate"].isnull().sum())

print(df.nunique())
print(df_clean.nunique())
# print(df.value_counts())
# print(df_clean.value_counts())
df_clean['Age']
# print(df_clean.duplicated().sum())

CustomerID               1500
Gender                      9
Age                        78
Region                     13
MembershipLevel            10
VisitCount                 26
AveragePurchaseAmount    1022
TotalPurchaseAmount      1377
SatisfactionScore          38
LastPurchaseDate          251
CouponUsed                  6
PreferredCategory          12
SignupDate               1065
Email                    1500
PurchaseStatus              2
dtype: int64
CustomerID               1500
Gender                      3
Age                        62
Region                      7
MembershipLevel             4
VisitCount                 26
AveragePurchaseAmount    1014
TotalPurchaseAmount      1377
SatisfactionScore          38
LastPurchaseDate          220
CouponUsed                  2
PreferredCategory           6
SignupDate               1056
Email                    1500
PurchaseStatus              2
dtype: int64


0        40.0
1        39.0
2       105.0
3        44.0
4        37.0
        ...  
1555     47.0
1556     30.0
1557     32.0
1558     45.0
1559     45.0
Name: Age, Length: 1560, dtype: float64

In [6]:
# TODO: 결측치와 전체 행 기준 중복 데이터를 처리하세요.

df_clean = df_clean.drop_duplicates().reset_index(drop=True)

df_clean["Gender"] = df_clean["Gender"].fillna("Unknown")
print(df_clean["Gender"].value_counts(dropna=False))

df_other = df_clean[df_clean["Gender"] == "Other"]
print("Other 고객:", len(df_other), "명, 구매율", round(df_other["PurchaseStatus"].mean(), 3))

#나이 100이상// #수치형은 중앙값, 범주형은 최빈값
df_clean.loc[df_clean['Age'] > 100, 'Age'] = np.nan
df_clean['Age'] = df_clean['Age'].fillna(df_clean['Age'].median())
df_clean['SatisfactionScore'] = df_clean['SatisfactionScore'].fillna(df_clean['SatisfactionScore'].median())
df_clean['AveragePurchaseAmount'] = df_clean['AveragePurchaseAmount'].fillna(df_clean['AveragePurchaseAmount'].median())
df_clean['Region'] = df_clean['Region'].fillna(df_clean['Region'].mode()[0])
df_clean['PreferredCategory'] = df_clean['PreferredCategory'].fillna(df_clean['PreferredCategory'].mode()[0])
df_clean['MembershipLevel'] = df_clean['MembershipLevel'].fillna(df_clean['MembershipLevel'].mode()[0])
median_date = df_clean["LastPurchaseDate"].median()
df_clean["LastPurchaseDate"] = df_clean["LastPurchaseDate"].fillna(median_date)

# TODO: 처리 전후 결과를 확인하세요.
print(f"처리 전 행 수: {len(df)}")
print(f"처리 후 행 수: {len(df_clean)}")
print("중복:", df_clean.duplicated().sum(), "건")
print("결측:", df_clean.isnull().sum()) #결측치
# # 제거된 중복 행 수
# print(f"제거된 중복 행 수: {len(df) - len(df_clean)}")


# print("===== 최종 데이터 정보 =====")
# df_clean.info()

# print("\n===== 결측치 =====")
# print(df_clean.isna().sum())

# print("\n===== 중복 데이터 =====")
# print(df_clean.duplicated().sum())

# print("\n===== 데이터 타입 =====")
# print(df_clean.dtypes)

Gender
Female     731
Male       696
Other       47
Unknown     26
Name: count, dtype: int64
Other 고객: 47 명, 구매율 0.596
처리 전 행 수: 1560
처리 후 행 수: 1500
중복: 0 건
결측: CustomerID               0
Gender                   0
Age                      0
Region                   0
MembershipLevel          0
VisitCount               0
AveragePurchaseAmount    0
TotalPurchaseAmount      0
SatisfactionScore        0
LastPurchaseDate         0
CouponUsed               0
PreferredCategory        0
SignupDate               0
Email                    0
PurchaseStatus           0
dtype: int64


In [41]:
# print(df_clean["Region"].mode())
# print(df_clean["PreferredCategory"].mode())
# print(df_clean["MembershipLevel"].mode())
# print(df_clean["AveragePurchaseAmount"].median())
# print(df_clean["SatisfactionScore"].median())
# print(df_clean["LastPurchaseDate"].median()) --결측값 처리하면 구매날짜 만들어짐

In [7]:
# def get_iqr_bounds(series):
    # TODO: Q1, Q3, IQR, 하한과 상한을 반환하세요. --> return
def get_iqr_bounds(series):
    q1 = series.quantile(0.25)
    q3 = series.quantile(0.75)
    iqr = q3 - q1
    lower = q1 - 1.5 * iqr
    upper = q3 + 1.5 * iqr
    return q1, q3, iqr, lower, upper

# TODO: 주요 수치형 컬럼의 이상치를 탐지하고 처리하세요.
age_q1 = df_clean["Age"].quantile(0.25)
age_q3 = df_clean["Age"].quantile(0.75)
age_iqr = age_q3 - age_q1
print("Q1:", age_q1)
print("Q3:", age_q3)
print("IQR:", age_iqr)
age_lower = age_q1 - 1.5 * age_iqr
age_upper = age_q3 + 1.5 * age_iqr
print("하한:", age_lower)
print("상한:", age_upper)
age_outliers = (df_clean["Age"] < age_lower) | (df_clean["Age"] > age_upper)
age_before_count = age_outliers.sum()
print("이상치 후보 수:", age_before_count)
df_clean.loc[age_outliers, ["CustomerID", "Age"]].head()

visitcount_q1 = df_clean["VisitCount"].quantile(0.25)
visitcount_q3 = df_clean["VisitCount"].quantile(0.75)
visitcount_iqr = visitcount_q3 - visitcount_q1
print("Q1:", visitcount_q1)
print("Q3:", visitcount_q3)
print("IQR:", visitcount_iqr)
visitcount_lower = visitcount_q1 - 1.5 * visitcount_iqr
visitcount_upper = visitcount_q3 + 1.5 * visitcount_iqr
print("하한:", visitcount_lower)
print("상한:", visitcount_upper)
visitcount_outliers = (df_clean["VisitCount"] < visitcount_lower) | (df_clean["VisitCount"] > visitcount_upper)
visitcount_before_count = visitcount_outliers.sum()
print("이상치 후보 수:", visitcount_before_count)
df_clean.loc[visitcount_outliers, ["CustomerID", "VisitCount"]].head()

avg_pur_amount_q1 = df_clean["AveragePurchaseAmount"].quantile(0.25)
avg_pur_amount_q3 = df_clean["AveragePurchaseAmount"].quantile(0.75)
avg_pur_amount_iqr = avg_pur_amount_q3 - avg_pur_amount_q1
print("Q1:", avg_pur_amount_q1)
print("Q3:", avg_pur_amount_q3)
print("IQR:", avg_pur_amount_iqr)
avg_pur_amount_lower = avg_pur_amount_q1 - 1.5 * avg_pur_amount_iqr
avg_pur_amount_upper = avg_pur_amount_q3 + 1.5 * avg_pur_amount_iqr
print("하한:", avg_pur_amount_lower)
print("상한:", avg_pur_amount_upper)
avg_pur_amount_outliers = (df_clean["AveragePurchaseAmount"] < avg_pur_amount_lower) | (df_clean["AveragePurchaseAmount"] > avg_pur_amount_upper)
avg_pur_amount_before_count = avg_pur_amount_outliers.sum()
print("이상치 후보 수:", avg_pur_amount_before_count)
df_clean.loc[avg_pur_amount_outliers, ["CustomerID", "AveragePurchaseAmount"]].head()

avg_pur_amount_q1 = df_clean["AveragePurchaseAmount"].quantile(0.25)
avg_pur_amount_q3 = df_clean["AveragePurchaseAmount"].quantile(0.75)
avg_pur_amount_iqr = avg_pur_amount_q3 - avg_pur_amount_q1
print("Q1:", avg_pur_amount_q1)
print("Q3:", avg_pur_amount_q3)
print("IQR:", avg_pur_amount_iqr)
avg_pur_amount_lower = avg_pur_amount_q1 - 1.5 * avg_pur_amount_iqr
avg_pur_amount_upper = avg_pur_amount_q3 + 1.5 * avg_pur_amount_iqr
print("하한:", avg_pur_amount_lower)
print("상한:", avg_pur_amount_upper)
avg_pur_amount_outliers = (df_clean["AveragePurchaseAmount"] < avg_pur_amount_lower) | (df_clean["AveragePurchaseAmount"] > avg_pur_amount_upper)
avg_pur_amount_before_count = avg_pur_amount_outliers.sum()
print("이상치 후보 수:", avg_pur_amount_before_count)
df_clean.loc[avg_pur_amount_outliers, ["CustomerID", "AveragePurchaseAmount"]].head()

to_amount_q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
to_amount_q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
to_amount_iqr = to_amount_q3 - to_amount_q1
print("Q1:", to_amount_q1)
print("Q3:", to_amount_q3)
print("IQR:", to_amount_iqr)
to_amount_lower = to_amount_q1 - 1.5 * to_amount_iqr
to_amount_upper = to_amount_q3 + 1.5 * to_amount_iqr
print("하한:", to_amount_lower)
print("상한:", to_amount_upper)
to_amount_outliers = (df_clean["TotalPurchaseAmount"] < to_amount_lower) | (df_clean["TotalPurchaseAmount"] > to_amount_upper)
to_amount_before_count = to_amount_outliers.sum()
print("이상치 후보 수:", to_amount_before_count)
df_clean.loc[to_amount_outliers, ["CustomerID", "TotalPurchaseAmount"]].head()

satis_q1 = df_clean["SatisfactionScore"].quantile(0.25)
satis_q3 = df_clean["SatisfactionScore"].quantile(0.75)
satis_iqr = satis_q3 - satis_q1
print("Q1:", satis_q1)
print("Q3:", satis_q3)
print("IQR:", satis_iqr)
satis_lower = satis_q1 - 1.5 * satis_iqr
satis_upper = satis_q3 + 1.5 * satis_iqr
print("하한:", satis_lower)
print("상한:", satis_upper)
satis_outliers = (df_clean["SatisfactionScore"] < satis_lower) | (df_clean["SatisfactionScore"] > satis_upper)
satis_before_count = satis_outliers.sum()
print("이상치 후보 수:", satis_before_count)
df_clean.loc[satis_outliers, ["CustomerID", "SatisfactionScore"]].head()

# TODO: 처리 전후 이상치 상태를 확인하세요.
df_before_outlier = df_clean.copy()

df_clean["Age"] = df_clean["Age"].astype("float64")
df_clean["Age"] = df_clean["Age"].clip(lower=age_lower, upper=age_upper)
changed = df_before_outlier["Age"] != df_clean["Age"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "Age"]
comparison["처리 후"] = df_clean.loc[changed, "Age"]
comparison.head()
df_clean["Age"].value_counts(dropna=False)

df_clean["VisitCount"] = df_clean["VisitCount"].astype("float64")
df_clean["VisitCount"] = df_clean["VisitCount"].clip(lower=visitcount_lower, upper=visitcount_upper)
changed = df_before_outlier["VisitCount"] != df_clean["VisitCount"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "VisitCount"]
comparison["처리 후"] = df_clean.loc[changed, "VisitCount"]
comparison.head()
df_clean["VisitCount"].value_counts(dropna=False)

df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].astype("float64")
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].clip(lower=avg_pur_amount_lower, upper=avg_pur_amount_upper)
changed = df_before_outlier["AveragePurchaseAmount"] != df_clean["AveragePurchaseAmount"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "AveragePurchaseAmount"]
comparison["처리 후"] = df_clean.loc[changed, "AveragePurchaseAmount"]
comparison.head()
df_clean["AveragePurchaseAmount"].value_counts(dropna=False)

df_clean["TotalPurchaseAmount"] = df_clean["TotalPurchaseAmount"].astype("float64")
df_clean["TotalPurchaseAmount"] = df_clean["TotalPurchaseAmount"].clip(lower=to_amount_lower, upper=to_amount_upper)
changed = df_before_outlier["TotalPurchaseAmount"] != df_clean["TotalPurchaseAmount"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "TotalPurchaseAmount"]
comparison["처리 후"] = df_clean.loc[changed, "TotalPurchaseAmount"]
comparison.head()
df_clean["TotalPurchaseAmount"].value_counts(dropna=False)

df_clean["SatisfactionScore"] = df_clean["SatisfactionScore"].astype("float64")
df_clean["SatisfactionScore"] = df_clean["SatisfactionScore"].clip(lower=satis_lower, upper=satis_upper)
changed = df_before_outlier["SatisfactionScore"] != df_clean["SatisfactionScore"]
print("변경한 값 수:", changed.sum())
comparison = pd.DataFrame()
comparison["처리 전"] = df_before_outlier.loc[changed, "SatisfactionScore"]
comparison["처리 후"] = df_clean.loc[changed, "SatisfactionScore"]
comparison.head()
df_clean["SatisfactionScore"].value_counts(dropna=False)


Q1: 32.0
Q3: 47.0
IQR: 15.0
하한: 9.5
상한: 69.5
이상치 후보 수: 6
Q1: 7.0
Q3: 11.0
IQR: 4.0
하한: 1.0
상한: 17.0
이상치 후보 수: 15
Q1: 83700.0
Q3: 150450.0
IQR: 66750.0
하한: -16425.0
상한: 250575.0
이상치 후보 수: 21
Q1: 83700.0
Q3: 150450.0
IQR: 66750.0
하한: -16425.0
상한: 250575.0
이상치 후보 수: 21
Q1: 186425.0
Q3: 536025.0
IQR: 349600.0
하한: -337975.0
상한: 1060425.0
이상치 후보 수: 53
Q1: 3.2
Q3: 4.2
IQR: 1.0
하한: 1.7000000000000002
상한: 5.7
이상치 후보 수: 10
변경한 값 수: 6
변경한 값 수: 15
변경한 값 수: 21
변경한 값 수: 53
변경한 값 수: 10


SatisfactionScore
3.7    107
3.6     80
3.9     77
3.3     74
3.8     70
3.5     69
4.2     68
3.2     64
4.1     63
5.0     63
3.4     62
4.0     58
3.0     57
3.1     56
4.4     52
4.3     48
4.5     47
4.7     38
4.6     37
2.7     35
2.8     35
2.6     33
4.9     33
2.9     32
4.8     25
2.4     25
2.5     19
2.3     19
2.2     14
2.1     11
1.7     10
2.0     10
1.8      6
1.9      3
Name: count, dtype: int64

In [8]:
num_cols = ["Age", "VisitCount", "AveragePurchaseAmount", "TotalPurchaseAmount", "SatisfactionScore"]

# ── 처리 전 이상치 탐지 ──
print("[이상치 처리 전]")
for col in num_cols:
    q1, q3, iqr, lower, upper = get_iqr_bounds(df_clean[col])
    n_out = ((df_clean[col] < lower) | (df_clean[col] > upper)).sum()
    print(f"{col}: 하한 {lower:,.1f} / 상한 {upper:,.1f} / 이상치 후보 {n_out}건")


outliers = df_clean[(df_clean['TotalPurchaseAmount'] < lower) | (df_clean['TotalPurchaseAmount'] > upper)]
print(f'IQR 범위: {lower:.2f} ~ {upper:.2f}')
print(f'이상치:\n{outliers}')

[이상치 처리 전]
Age: 하한 9.5 / 상한 69.5 / 이상치 후보 0건
VisitCount: 하한 1.0 / 상한 17.0 / 이상치 후보 0건
AveragePurchaseAmount: 하한 -16,425.0 / 상한 250,575.0 / 이상치 후보 0건
TotalPurchaseAmount: 하한 -337,975.0 / 상한 1,060,425.0 / 이상치 후보 0건
SatisfactionScore: 하한 1.7 / 상한 5.7 / 이상치 후보 0건
IQR 범위: 1.70 ~ 5.70
이상치:
      CustomerID  Gender   Age    Region MembershipLevel  VisitCount  \
0     CUST101486  Female  40.0     Seoul          Silver         5.0   
1     CUST100113    Male  39.0  Gyeonggi           Basic         7.0   
2     CUST100168    Male  39.0   Gwangju          Silver        12.0   
3     CUST101212    Male  44.0     Daegu             VIP         7.0   
4     CUST100312    Male  37.0  Gyeonggi           Basic        14.0   
...          ...     ...   ...       ...             ...         ...   
1495  CUST101044  Female  47.0   Daejeon            Gold         6.0   
1496  CUST100737    Male  47.0     Seoul           Basic        14.0   
1497  CUST100506    Male  30.0     Busan          Silver         8.

In [9]:
q1 = df_clean["Age"].quantile(0.25)
q3 = df_clean["Age"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["Age"] < lower) | (df_clean["Age"] > upper)
print("age 처리 전 후보 수:", age_before_count)
print("age 처리 후 후보 수:", outliers.sum())

q1 = df_clean["VisitCount"].quantile(0.25)
q3 = df_clean["VisitCount"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["VisitCount"] < lower) | (df_clean["VisitCount"] > upper)
print("visit 처리 전 후보 수:", visitcount_before_count)
print("visit 처리 후 후보 수:", outliers.sum())

q1 = df_clean["AveragePurchaseAmount"].quantile(0.25)
q3 = df_clean["AveragePurchaseAmount"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["AveragePurchaseAmount"] < lower) | (df_clean["AveragePurchaseAmount"] > upper)
print("avg 처리 전 후보 수:", avg_pur_amount_before_count)
print("avg 처리 후 후보 수:", outliers.sum())


q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["TotalPurchaseAmount"] < lower) | (df_clean["TotalPurchaseAmount"] > upper)
print("total 처리 전 후보 수:", to_amount_before_count)
print("total 처리 후 후보 수:", outliers.sum())

q1 = df_clean["SatisfactionScore"].quantile(0.25)
q3 = df_clean["SatisfactionScore"].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr
outliers = (df_clean["SatisfactionScore"] < lower) | (df_clean["SatisfactionScore"] > upper)
print("satis 처리 전 후보 수:", satis_before_count)
print("satis 처리 후 후보 수:", outliers.sum())

age 처리 전 후보 수: 6
age 처리 후 후보 수: 0
visit 처리 전 후보 수: 15
visit 처리 후 후보 수: 0
avg 처리 전 후보 수: 21
avg 처리 후 후보 수: 0
total 처리 전 후보 수: 53
total 처리 후 후보 수: 0
satis 처리 전 후보 수: 10
satis 처리 후 후보 수: 0


## 문제 3. 분석용 변수 생성 및 다음 단계 데이터 준비

### 업무 상황

정제된 데이터를 CRM팀의 고객군 분석과 구매 예측 업무에 활용하려면, 원본 컬럼만으로 확인하기 어려운 고객 특성을 분석용 변수로 표현해야 합니다.  
필요한 고객을 조건에 따라 추출하고, 이후 Part 2와 Part 3에서 활용할 파생변수를 만든 뒤 최종 데이터를 저장하세요.

### 요구사항

1. 정제된 데이터에서 업무적으로 의미 있는 조건을 설정하여 고객 데이터를 추출하세요.
2. 고객 분석에 활용할 파생변수를 **2개 이상** 생성하세요.
3. 생성한 파생변수의 값과 분포가 적절한지 확인하세요.
4. 최종 데이터의 결측치, 중복값, 고객 식별자와 타깃 값을 검증하세요.
5. 전처리 완료 데이터를 `전자상거래_고객구매_전처리완료.csv`로 저장하세요.

### 힌트

- 연령대, 구매 수준, 최근 구매 여부, 우수 고객 여부 등을 파생변수 후보로 고려할 수 있습니다.
- 조건 추출 결과는 별도 DataFrame으로 확인해도 되며, 최종 데이터 전체를 삭제할 필요는 없습니다.

- 고객 연령을 구간화한 `AgeGroup`을 생성하세요.
- 총구매금액을 기준으로 `PurchaseGrade`를 생성하세요.

In [10]:
REFERENCE_DATE = pd.Timestamp("2026-06-01")

# TODO: 업무적으로 의미 있는 조건을 설정하여 고객 데이터를 추출하세요.
visit_median = df_clean["VisitCount"].median()
visit_condition = df_clean["VisitCount"] >= visit_median
purchase_condition = df_clean["TotalPurchaseAmount"] > 0
selected_customers = df_clean[visit_condition & purchase_condition]
selected_customers = selected_customers.copy()
print("조건에 맞는 고객 수:", len(selected_customers))
selected_customers.head()
# TODO: 분석용 파생변수를 2개 이상 생성하세요. // 나이 구간별 cut // 구매금액 low - mediun - high
age_bins = [0, 29, 39, 49, 59, float("inf")]
age_labels = ["20대 이하", "30대", "40대", "50대", "60대 이상"]
df_clean["AgeGroup"] = pd.cut(
    df_clean["Age"],
    bins=age_bins,
    labels=age_labels,
    include_lowest=True
)
display(df_clean[["Age", "AgeGroup"]].head(10))

purchase_q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
purchase_q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
print("Q1:", purchase_q1)
print("Q3:", purchase_q3)
df_clean["PurchaseGrade"] = "Medium"
low_condition = df_clean["TotalPurchaseAmount"] <= purchase_q1
df_clean.loc[low_condition, "PurchaseGrade"] = "Low"
high_condition = df_clean["TotalPurchaseAmount"] > purchase_q3
df_clean.loc[high_condition, "PurchaseGrade"] = "High"
df_clean["PurchaseGrade"].value_counts(dropna=False)
df_clean[["TotalPurchaseAmount", "PurchaseGrade"]].head(10)

# TODO: 파생변수의 값과 분포를 확인하세요.
df_clean["AgeGroup"].value_counts(dropna=False)
df_clean["AgeGroup"].isna().sum()
df_clean["PurchaseGrade"].value_counts(dropna=False)
df_clean["PurchaseGrade"].isna().sum()
df_clean.groupby("PurchaseGrade")["TotalPurchaseAmount"].min() #등급별 구매급액 범위
df_clean.groupby("PurchaseGrade")["TotalPurchaseAmount"].max()
df_clean.groupby("PurchaseGrade")["TotalPurchaseAmount"].median()

example_ages = pd.DataFrame()
example_ages["예시 나이"] = [19, 20, 29, 30, 39, 40, 59, 60]
example_ages["구간"] = pd.cut(example_ages["예시 나이"], bins=age_bins, labels=age_labels, include_lowest=True)
example_ages

middle_amount = (purchase_q1 + purchase_q3) / 2
example_amounts = pd.DataFrame()
example_amounts["예시 금액"] = [purchase_q1, middle_amount, purchase_q3, purchase_q3 + 1]
example_amounts["구매등급"] = "Medium"
low_example = example_amounts["예시 금액"] <= purchase_q1
example_amounts.loc[low_example, "구매등급"] = "Low"
high_example = example_amounts["예시 금액"] > purchase_q3
example_amounts.loc[high_example, "구매등급"] = "High"
example_amounts

조건에 맞는 고객 수: 831


,Age,AgeGroup
0,40.0,40대
1,39.0,30대
2,39.0,30대
3,44.0,40대
4,37.0,30대
5,45.0,40대
6,47.0,40대
7,26.0,20대 이하
8,56.0,50대
9,42.0,40대


Q1: 186425.0
Q3: 536025.0


,예시 금액,구매등급
0,186425.0,Low
1,361225.0,Medium
2,536025.0,Medium
3,536026.0,High


In [11]:
OUTPUT_PATH = Path("전자상거래_고객구매_전처리완료.csv")

# TODO: 최종 데이터 품질을 검증하세요.
missing_count = df_clean.isna().sum().sum()
print("전체 결측 수 (0이어야 함):", missing_count)
duplicate_count = df_clean.duplicated().sum()
print("전체 행 중복 수 (0이어야 함):", duplicate_count)
id_duplicate_count = df_clean["CustomerID"].duplicated().sum()
id_missing_count = df_clean["CustomerID"].isna().sum()
print("ID 중복 수:", id_duplicate_count)
print("ID 결측 수:", id_missing_count)
id_text = df_clean["CustomerID"].astype("string")
id_text = id_text.str.strip()
id_empty_count = (id_text == "").sum()
print("공백만 있는 ID 수 (0이어야 함):", id_empty_count)
df_clean["PurchaseStatus"].value_counts(dropna=False)
target_valid = df_clean["PurchaseStatus"].isin([0, 1]).all()
print("타깃이 모두 0 또는 1인가 (True여야 함):", target_valid)
print("AgeGroup" in df_clean.columns)
print("PurchaseGrade" in df_clean.columns)

assert missing_count == 0, "결측치가 남아 있습니다."
assert duplicate_count == 0, "전체 행 중복이 남아 있습니다."
assert id_duplicate_count == 0, "고객 ID가 중복됩니다."
assert id_missing_count == 0, "고객 ID가 비어 있습니다."
assert id_empty_count == 0, "공백만 있는 고객 ID가 있습니다."
assert target_valid, "타깃에 0/1 이외의 값이 있습니다."
assert "AgeGroup" in df_clean.columns
assert "PurchaseGrade" in df_clean.columns
print("최종 품질 검증 통과:", df_clean.shape)
# TODO: 전처리 완료 데이터를 CSV 파일로 저장하세요.
df_to_save = df_clean.copy()
df_to_save["LastPurchaseDate"] = df_to_save["LastPurchaseDate"].dt.strftime("%Y-%m-%d")
df_to_save["SignupDate"] = df_to_save["SignupDate"].dt.strftime("%Y-%m-%d")
df_to_save.to_csv("전자상거래_고객구매_전처리완료.csv", index=False, encoding="utf-8-sig")
saved_df = pd.read_csv("전자상거래_고객구매_전처리완료.csv")
saved_df.head()
print("정제 데이터 크기:", df_clean.shape)
print("다시 읽은 크기:", saved_df.shape)
print("컬럼 일치:", saved_df.columns.equals(df_clean.columns))
saved_df.isna().sum()

assert saved_df.shape == df_clean.shape
assert saved_df.columns.equals(df_clean.columns)
assert saved_df.isna().sum().sum() == 0
assert saved_df["CustomerID"].is_unique
print("CSV 저장 및 다시 읽기 확인 완료")


전체 결측 수 (0이어야 함): 0
전체 행 중복 수 (0이어야 함): 0
ID 중복 수: 0
ID 결측 수: 0
공백만 있는 ID 수 (0이어야 함): 0
타깃이 모두 0 또는 1인가 (True여야 함): True
True
True
최종 품질 검증 통과: (1500, 17)
정제 데이터 크기: (1500, 17)
다시 읽은 크기: (1500, 17)
컬럼 일치: True
CSV 저장 및 다시 읽기 확인 완료


# 문제 4. 데이터 정제 및 파생변수 설계 근거 설명

## 업무 상황

CRM팀은 전처리 결과가 단순히 실행되는 것뿐 아니라,
적용한 처리 기준과 파생변수가 실제 분석 목적에 적합한지 확인하려고 합니다.

문제 1~3에서 수행한 결과를 바탕으로 다음 내용을 설명하세요.

## 요구사항

### 4-1. 데이터 품질 문제의 영향과 한계

1. 문제 1에서 확인한 주요 품질 문제를 한 가지 이상 선택하세요.
2. 해당 문제가 이후 EDA 또는 모델링 결과에 미칠 수 있는 영향을 설명하세요.
3. 현재 데이터만으로 판단하기 어려운 점이나 추가 확인이 필요한 조건을 작성하세요.

### 4-2. 결측치·이상치 처리 방법의 선택 근거

1. 적용한 결측치 처리 방법을 한 가지 이상 제시하세요.
2. 해당 방법을 선택한 이유를 데이터 특성과 연결하여 작성하세요.
3. 적용한 이상치 처리 방법과 처리 시 주의할 점을 설명하세요.

### 4-3. 파생변수의 활용 의도와 한계

1. `파생변수1`과 `파생변수2`의 생성 기준을 설명하세요.
2. 각 파생변수가 Part 2 고객 특성 분석에서 어떻게 활용될 수 있는지 작성하세요.
3. 파생변수 사용 시 발생할 수 있는 정보 손실 또는 해석상의 한계를 설명하세요.

## 문제 4 결과 작성

### 4-1. 데이터 품질 문제의 영향과 한계

- 선택한 품질 문제:
-> Age에서 34세, thirty, ?, unknown, 빈칸과 같은 문자열이 섞여 있음을 확인했다
-> Gender에서 Male과 M, male이 섞여 있어 같은 성별이 다른 범주로 저장되어 있었다.

- EDA 또는 모델링에 미치는 영향:
-> Age가 문자열이면 평균이나 계산을 할 수 없고, 숫자 크기를 확인할 수 없다.
-> 같은 성별이 다른 범주로 저장되어 여러 범주로 나뉘어지는데 실제 고객 수에 오류가 발생한다.

- 추가 확인이 필요한 조건 또는 한계:
-> Age가 100세 이상이 실제 고령의 고객인지 입력 오류인지 수치만으로 확실하지가 않다.(특히 130)
다만 온라인 쇼핑몰에서 100세 이상의 구매고객은 현실성이 없다고 판단해 오류로 처리했다.
-> Gender에 Other가 선택사항에 있었는지, 입력오류인지 데이터만으로 판단하기에 확실하지 않다. Unknown도 입력 선택이었는지, 오류인지, 비회원 주문인지 확인이 필요하다.

### 4-2. 결측치·이상치 처리 방법의 선택 근거

- 결측치 처리 방법:
-> Age의 결측은 중앙값로 대체했다.(빈칸 50)
-> Gender의 결측은 Unknown이라는 값으로 대체했다.(빈칸 26)
- 해당 방법을 선택한 이유:
-> Age : 130세 같이 큰 값으로 인해 평균이 올라가는데 중앙값은 이런 극단값의 영향을 평균보다 덜 받는다. 다만 빈칸 모두 중앙값으로 대체해 실제 나이를 알아낸 것은 아니다.
-> Gender : 여자 731명, 남자 696명으로 결측값을 어느 성별로 정하기엔 숫자의 차이가 크지 않았다. 그래서 Unknown(26)으로 따로 두었다. 또한 Other(47)는 성별 선택지 중의 하나로 만들었다.
- 이상치 처리 방법: IQR로 이상치 후보를 찾았다. 상한을 넘거나 하한보다 작으면 이상치로 간주했다. Age는 상한이 69.5세였고 100세 넘는 구매건은 잘못 입력된 값으로 봤다. 빈칸으로 처리 후 중앙값으로 대체했다.
- 이상치 처리 시 주의점: 상한보다 더 높은 값, 하한보다 더 낮은 값을 최빈값, 혹은 중앙값으로 대체한 것이라 이상치 처리 후 무조건 지우면 안된다.

### 4-3. 파생변수의 활용 의도와 한계

- `AgeGroup` 생성 기준과 활용 목적: Age를 0~29, 30~39, 40~49, 50~59, 60 이상으로 나누고
   20대 이하·30대·40대·50대·60대 이상으로 분류했다. 연령대별로 고객수, 평균 방문횟수를 비교해 집단별 이용특성을 살펴볼 수 있다.
- `PurchaseGrade` 생성 기준과 활용 목적: 총구매금액을 기준으로 Q1 186,425원 이하는 Low, 536,025원 Q3 초과는 High, Q1 초과 Q3 536,025원 이하를 Medium으로 정했다. 구매등급별 만족도나 선호 카테고리를 비교하여 등급별 고객 특성을 살펴볼 수 있다.
  
- 파생변수 사용 시 한계: 
->20세와 29세가 똑같이 20대 이하가 되고,
   50만원 쓴 사람과 100만원 쓴 사람이 똑같이 High가 된다. 나이가 차이나도, 금액이 차이나도 같은 구간으로 묶여있으면 차이가 사라진다. 반대로 Q1 기준금액이 186,425원으로 186,200원과 186,500원은 300원 차이인데 등급이 갈린다.
->Age 결측을 중앙값인 39세로 채웠는데, 39세는 30대 구간의 끝이다.
   그래서 30대가 521명으로 가장 많아진 데에 영향이 있어 AgeGroup도 실제 연령대와 다를 수 있다.
-> 분위수 Q1, Q3는 데이터가 바뀌면 같이 달라져 기간별 등급을 바로 비교하기 어려울 수 있다.

In [12]:
age_bins = [0, 29, 39, 49, 59, float("inf")]
age_labels = ["20대 이하", "30대", "40대", "50대", "60대 이상"]
df_clean["AgeGroup"] = pd.cut(
    df_clean["Age"],
    bins=age_bins,
    labels=age_labels,
    include_lowest=True
)
display(df_clean[["Age", "AgeGroup"]].head(10))

,Age,AgeGroup
0,40.0,40대
1,39.0,30대
2,39.0,30대
3,44.0,40대
4,37.0,30대
5,45.0,40대
6,47.0,40대
7,26.0,20대 이하
8,56.0,50대
9,42.0,40대


In [13]:
purchase_q1 = df_clean["TotalPurchaseAmount"].quantile(0.25)
purchase_q3 = df_clean["TotalPurchaseAmount"].quantile(0.75)
print("Q1:", purchase_q1)
print("Q3:", purchase_q3)
df_clean["PurchaseGrade"] = "Medium"
low_condition = df_clean["TotalPurchaseAmount"] <= purchase_q1
df_clean.loc[low_condition, "PurchaseGrade"] = "Low"
high_condition = df_clean["TotalPurchaseAmount"] > purchase_q3
df_clean.loc[high_condition, "PurchaseGrade"] = "High"
df_clean["PurchaseGrade"].value_counts(dropna=False)
df_clean[["TotalPurchaseAmount", "PurchaseGrade"]].head(10)

Q1: 186425.0
Q3: 536025.0


,TotalPurchaseAmount,PurchaseGrade
0,139800.0,Low
1,159200.0,Low
2,1055800.0,High
3,413200.0,Medium
4,576000.0,High
5,123000.0,Low
6,1060425.0,High
7,40700.0,Low
8,443500.0,Medium
9,70700.0,Low
